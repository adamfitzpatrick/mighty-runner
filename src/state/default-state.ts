import { Character, PersonalData, CharacterIdentifier, Attributes, Effect, GearItem, Pic } from '@models'

export interface AppState {
  characters: Character[] | null
  activeCharacter: CharacterIdentifier | null
  pic: Pic | null
  personalData: PersonalData | null
  attributes: Attributes | null
  gear: GearItem[] | null
  effects: Effect[] | null
  apiError: boolean
}

export const defaultState: AppState = {
  characters: null,
  pic: null,
  activeCharacter: null,
  personalData: null,
  attributes: null,
  gear: null,
  effects: null,
  apiError: false
}

export default defaultState
