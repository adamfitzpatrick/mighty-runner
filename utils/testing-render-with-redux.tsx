import * as React from 'react'
import {
  createStore,
  combineReducers,
  AnyAction,
  Store,
  Dispatch,
  applyMiddleware
} from 'redux'
import { Provider } from 'react-redux'
import { render, RenderResult } from '@testing-library/react'
import * as Models from '@models'
import { AppState } from '../src/state/default-state'

export type RenderWithRedux =
  RenderResult &
  {
    store: Store<AppState, AnyAction>,
    dispatchSpy: jest.Mock<Dispatch<AnyAction>>
  }

const reducer = combineReducers({
  characters: (state: Models.Character[] | null = null) => state,
  activeCharacter: (state: Models.CharacterIdentifier | null = null) => state,
  personalData: (state: Models.PersonalData | null = null) => state,
  attributes: (state: Models.Attributes | null = null) => state,
  gear: (state: Models.GearItem[] | null = null) => state,
  effects: (state: Models.Effect[] | null = null) => state
})

const getSpyMiddleware = (dispatchSpy: jest.Mock<Dispatch<AnyAction>>) => () => {
  return (next: Dispatch<AnyAction>) => (action: AnyAction) => {
    dispatchSpy(action)
    next(action)
  }
}

export function renderWithRedux (
  ui: JSX.Element,
  initialState: AppState
): RenderWithRedux {
  const dispatchSpy: jest.Mock<Dispatch<AnyAction>> = jest.fn()
  const spyMiddleware = getSpyMiddleware(dispatchSpy)
  const store = createStore(reducer, initialState, applyMiddleware(spyMiddleware))

  return {
    ...render(<Provider store={store}>{ ui }</Provider>),
    store,
    dispatchSpy
  }
}
