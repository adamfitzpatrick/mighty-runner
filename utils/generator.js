const fs = require('fs')
const path = require('path')
const yargs = require('yargs')

if (!yargs.argv.container && !yargs.argv.component) {
  throw new Error('no component or container name provided')
}

const container = yargs.argv.container && yargs.argv.container.replace(/^[a-z]/, a => a.toUpperCase())
const component = yargs.argv.component && yargs.argv.component.replace(/^[a-z]/, a => a.toUpperCase())

let thing = yargs.argv.component
if (container) {
  thing = container
}
const lower = thing.replace(/^[A-Z]/, a => a.toLowerCase()).replace(/[A-Z]/g, a => `-${a.toLowerCase()}`)
thing = thing.replace(/^[a-z]/, a => a.toUpperCase()).replace(/-([A-Za-z0-9])/g, a => a.slice(1).toUpperCase())
scssClass = thing.replace(/^[A-Z]/, a => a.toLowerCase())

const thingTs = `import * as React from 'react'

import * as styles from './${lower}.scss'

interface Props {}

export default function ${thing} (props: Props) {
  return <div>${thing}</div>
}
`

const thingTestTs = `import * as React from 'react'
import { cleanup, render } from '@testing-library/react'
import { renderWithRedux, RenderWithRedux } from '../../../utils/testing-render-with-redux'
import ${thing} from '.'

describe('${thing} ${ container ? 'container' : 'component'}', () => {
  test('should do nothing', () => {
    expect(true).toEqual(true)
  })
})
`

const indexTs = `export { default } from './${lower}'
`

const scss = `.${scssClass} {
  display: block;
}
`

const scssD = `export const ${scssClass}: string
`

let dest = path.resolve(process.cwd(), 'src', 'components', lower)
if (container) {
  dest = path.resolve(process.cwd(), 'src', 'containers', lower)
}

fs.mkdirSync(dest)
fs.writeFileSync(path.resolve(dest, `${lower}.tsx`), thingTs)
fs.writeFileSync(path.resolve(dest, `${lower}.test.tsx`), thingTestTs)
fs.writeFileSync(path.resolve(dest, 'index.ts'), indexTs)
fs.writeFileSync(path.resolve(dest, `${lower}.scss`), scss)
fs.writeFileSync(path.resolve(dest, `${lower}.scss.d.ts`), scssD)
