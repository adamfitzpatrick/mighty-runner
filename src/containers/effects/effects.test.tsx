import * as React from 'react'
import { EffectsById, EffectsByTarget } from '.'
import { cleanup } from '@testing-library/react'
import { renderWithRedux, RenderWithRedux } from '../../../utils/testing-render-with-redux'
import { Effect } from '@models'
import { AppState } from '@state/default-state'
import { fireEvent } from '@testing-library/dom'

describe('Effects containers', () => {
  let effects: Effect[]
  let appState: AppState

  beforeEach(() => {
    effects = [{
      id: '1',
      name: 'a',
      description: 'd',
      target: [ 'a', 'b' ],
      active: false,
      value: 1
    }, {
      id: '2',
      name: 'b',
      description: 'd',
      target: [ 'a', 'c' ],
      active: true,
      value: 2
    }, {
      id: '3',
      name: 'c',
      description: 'd',
      target: [ 'a', 'b' ],
      active: false,
      value: 3
    }]
    appState = {
      effects
    } as AppState
  })

  describe('EffectsById', () => {
    let sut: RenderWithRedux

    beforeEach(() => {
      sut = renderWithRedux(<EffectsById ids={[ '1', '3' ]} />, appState)
    })

    it('should render correctly', () => {
      expect(sut.container.innerHTML).toMatchSnapshot()
      expect(sut.getAllByTestId('editable-effect.component').length).toBe(2)
    })

    it('should correctly handle missing data', () => {
      cleanup()
      sut = renderWithRedux(<EffectsById ids={[ '1' ]} />, { effects: null } as AppState)
      expect(sut.container.innerHTML).toMatchSnapshot()
    })
  })

  describe('EffectsByTarget', () => {
    let sut: RenderWithRedux

    beforeEach(() => {
      sut = renderWithRedux(<EffectsByTarget target={[ 'a', 'c' ]} />, appState)
    })

    it('should render correctly', () => {
      expect(sut.container.innerHTML).toMatchSnapshot()
      expect(sut.getAllByTestId('non-editable-effect.component').length).toBe(1)
    })

    it('should handle toggling the active state of an effect', () => {
      const input = sut.getAllByTestId('Active.input.component')[0]
      const event = {
        bubbles: true,
        target: {
          checked: true
        }
      }
      fireEvent(input, new MouseEvent('click', event))
      effects[1].active = false
      expect(sut.dispatchSpy).toHaveBeenCalledWith({
        type: 'UPDATE_EFFECT',
        payload: effects[1]
      })
    })

    it('should correctly handle missing data', () => {
      cleanup()
      sut = renderWithRedux(<EffectsByTarget target={[ 'a', 'b' ]} />, { effects: null } as AppState)
      expect(sut.container.innerHTML).toMatchSnapshot()
    })
  })
})
