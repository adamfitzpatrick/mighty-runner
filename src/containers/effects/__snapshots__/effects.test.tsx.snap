// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Effects containers EffectsById should render correctly when editing 1`] = `"<div data-testid=\\"editable-effect.component\\"><label><span>Name</span><input data-testid=\\"Name.input.component\\" type=\\"text\\" value=\\"a\\"></label><label><span>Description</span><input data-testid=\\"Description.input.component\\" type=\\"text\\" value=\\"d\\"></label><label>Target<input data-testid=\\"0.array.input\\" type=\\"text\\" value=\\"a\\"></label><label><span class=\\"additionalLabel\\">Target</span><input data-testid=\\"1.array.input\\" type=\\"text\\" value=\\"b\\"></label><label><span>Value</span><input data-testid=\\"Value.input.component\\" type=\\"text\\" value=\\"1\\"></label><label><span>Active</span><input data-testid=\\"Active.input.component\\" type=\\"checkbox\\" value=\\"\\"></label><button data-testid=\\"Delete Effect.button.component\\">Delete Effect</button></div><div data-testid=\\"editable-effect.component\\"><label><span>Name</span><input data-testid=\\"Name.input.component\\" type=\\"text\\" value=\\"c\\"></label><label><span>Description</span><input data-testid=\\"Description.input.component\\" type=\\"text\\" value=\\"d\\"></label><label>Target<input data-testid=\\"0.array.input\\" type=\\"text\\" value=\\"a\\"></label><label><span class=\\"additionalLabel\\">Target</span><input data-testid=\\"1.array.input\\" type=\\"text\\" value=\\"b\\"></label><label><span>Value</span><input data-testid=\\"Value.input.component\\" type=\\"text\\" value=\\"3\\"></label><label><span>Active</span><input data-testid=\\"Active.input.component\\" type=\\"checkbox\\" value=\\"\\"></label><button data-testid=\\"Delete Effect.button.component\\">Delete Effect</button></div>"`;
